### DISCUSSION TITLE
Equivalent of discriminated union with discriminator being part of the parent model

### DISCUSSION THREAD

**Discussion by anirban-ray-publicis-sapient** [pydantic/pydantic#5812] [Question]
Consider this example:

```python3
import typing

import pydantic


class Cat(pydantic.BaseModel):
    sub_category: typing.Literal["persian", "siamese"]
    name: str
    meows: int


class Dog(pydantic.BaseModel):
    sub_category: typing.Literal["labrador", "poodle"]
    name: str
    barks: int


class Pet(pydantic.BaseModel):
    category: typing.Literal["cat", "dog"]
    properties: Cat | Dog = pydantic.Field(..., discriminator="category")  # does not work, what to do


print(Pet(category="cat", properties={"sub_category": "siamese", "name": "luna", "meows": 1}))  # should pass
print(Pet(category="dog", properties={"sub_category": "labrador", "name": "charlie", "barks": 2}))  # should pass
print(Pet(category="cat", properties={"sub_category": "poodle", "name": "leo", "barks": 3}))  # should fail
print(Pet(category="dog", properties={"sub_category": "persian", "name": "felix", "meows": 4}))  # should fail

```

Here, I want to validate `properties` based on `category`. If `category` is `"cat"`, I want to check `properties` against `Cat`, and similarly for `"dog"` and `Dog`. Is this possible? In my problem, I cannot move the "discriminator" field ("category" here) in the union classes.

---

I can add a custom validator like this, but I think that'll validate first against all types in union which I'll prefer to avoid.

```python3
class Pet(pydantic.BaseModel):
    category: typing.Literal["cat", "dog"]
    properties: Cat | Dog

    @pydantic.validator("properties")
    def check_category_properties(
        cls: type["Pet"], passed_properties: Cat | Dog, values: dict
    ) -> Cat | Dog:
        del cls  # vulture

        if values["category"] == "cat":
            assert isinstance(passed_properties, Cat)
        elif values["category"] == "dog":
            assert isinstance(passed_properties, Dog)

        return passed_properties
```

**Comment by adriangb** [comment]
We don't currently support this, but I will note that SerDe does: https://serde.rs/enum-representations.html. So it's not an unreasonable feature request.

**Comment by youcandanch** [comment]
@adriangb dredging up some ancient history here! Is this still something that you'd see as fitting within the ethos of Pydantic? I'd be happy to take a crack at it, but didn't want to do so if it's a) already been done, or b) not something that'd fit into the current landscape. Thanks!

**Reply by adriangb** [reply]
I don't think anyone would be opposed to the feature on a philosophical level. The fact that SerDe has it is a good indication that it's not unreasonable to support. I would worry that it makes our internals more complicated or hurts performance in other cases by having to pass more information around. So you're welcome to take a crack at it, but don't be surprised if it's hard or impossible without major refactors which might not be accepted. I'll try to cook up an alternative that may not be pretty but at least should work.

**Reply by youcandanch** [reply]
Yeah, been reading the rough surface area of the code I think this would touch (which I am sure is like 50% of what it'd _actually_ touch) and agreed that it's a bit of a puzzler. Gonna sleep on it for a bit and try to do the same!

**Reply by youcandanch** [reply]
@adriangb poked around on this last night for a bit -- I think a potential way to do this while minimizing impact to the existing codebase would be to introduce an `AdjacentDiscriminator` class, which isn't _quite_ how SerDe does it, but _does_ mirror their terminology around it. I'm gonna take a stab at implementing that and if it looks viable, I'll put up a PR so we can talk the idea down.

**Comment by illeatmyhat** [comment]
+1. I need to be able to support declaring third party defined classes in config files. A concrete example would be something like:
```python
class Index(BaseModel):
    embedding_type: Literal['HuggingFaceEmbeddings', 'OpenAIEmbeddings']
    embedding: HuggingFaceEmbeddings | OpenAIEmbeddings = Field(external_discriminator='embedding_type')
    ...
```
Here, LangChain is a particularly practical example because it has many external contributors, and it seems like everything is a pydantic model.
I don't know if there is a better representation than this though.
